openssl-dynamic/src/main/c/cert_compress.c

```diff
@@ -0,0 +1,182 @@
+/*
+ * Copyright 2022 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+#include "tcn.h"
+#include "ssl_private.h"
+#ifdef OPENSSL_IS_BORINGSSL
+#include "cert_compress.h"
+
+static int compress(jobject compression_algorithm, jmethodID compress_method, SSL* ssl, CBB* out,
+    const uint8_t* in, size_t in_len) {
+
+    int ret = 0;
+    JNIEnv *e = NULL;
+    jbyteArray inputArray = NULL;
+
+    if (compression_algorithm == NULL || compress_method == NULL) {
+        goto complete;
+    }
+    if (tcn_get_java_env(&e) != JNI_OK) {
+        goto complete;
+    }
+    if ((inputArray = (*e)->NewByteArray(e, in_len)) == NULL) {
+        goto complete;
+    }
+
+    (*e)->SetByteArrayRegion(e, inputArray, 0, in_len, (jbyte*) in);
+
+    jbyteArray resultArray = (*e)->CallObjectMethod(e, compression_algorithm, compress_method,
+                    P2J(ssl), inputArray);
+
+    if ((*e)->ExceptionCheck(e) != JNI_FALSE) {
+        (*e)->ExceptionClear(e);
+        goto complete; // Exception while calling into Java
+    }
+    if (resultArray == NULL) {
+        goto complete; // Received NULL array from call to Java
+    }
+
+    int resultLen = (*e)->GetArrayLength(e, resultArray);
+    uint8_t* outData = NULL;
+    if (!CBB_reserve(out, &outData, resultLen)) {
+        goto complete; // Unable to reserve space for compressed data
+    }
+    jbyte* resultData = (*e)->GetByteArrayElements(e, resultArray, NULL);
+    memcpy(outData, resultData, resultLen);
+    if (!CBB_did_write(out, resultLen)) {
+        goto complete; // Unable to advance bytes written to CBB
+    }
+    (*e)->ReleaseByteArrayElements(e, resultArray, resultData, JNI_ABORT);
+    ret = 1; // Success
+    goto complete;
+
+complete:
+    // Free up any allocated memory and return.
+    if (inputArray != NULL) {
+        (*e)->DeleteLocalRef(e, inputArray);
```

Comment:
honestly I think we don't need to delete the local ref explicit.

Reply:
removed